scapy/layers/ipsec.py

```diff
@@ -912,12 +1202,19 @@ def _encrypt_esp(self, pkt, seq_num=None, iv=None, esn_en=None, esn=None):
         esp.data = payload
         esp.nh = nh
 
-        esp = self.crypt_algo.pad(esp)
-        esp = self.crypt_algo.encrypt(self, esp, self.crypt_key,
-                                      esn_en=esn_en or self.esn_en,
-                                      esn=esn or self.esn)
+        # Here we use either aead algo or crypt + auth
+        if self.aead_algo is None:
+            esp = self.crypt_algo.pad(esp)
+            esp = self.crypt_algo.encrypt(self, esp, self.crypt_key,
+                                          esn_en=esn_en or self.esn_en,
+                                          esn=esn or self.esn)
 
-        self.auth_algo.sign(esp, self.auth_key)
+            self.auth_algo.sign(esp, self.auth_key)
+        else:
+            esp = self.aead_algo.pad(esp)
+            esp = self.aead_algo.encrypt(self, esp, self.aead_key,
+                                         esn_en=esn_en or self.esn_en,
+                                         esn=esn or self.esn)
```

Comment:
Same this could probably have been unified

Reply:
Same comment.
In case of ESP we have the choice:
- Use Crypt + Auth
- Use AEAD

So in code I need to discriminate the choice. I choose to test if AEAD algo is None, the we use Crypt + Auth. If not None, we are using AEAD.